_posts/2025-02-28-a-practical-guide-for-selecting-HNSW-hyperparameters.md

```diff
@@ -0,0 +1,683 @@
+- **Search speed**, measured by query throughput---the number of queries executed per second.  
+
+While index build time and index size are also important, we leave those aspects for future discussion.  
+
+The structure of the HNSW graph is controlled by its hyperparameters, which determine how densely vectors are connected. A denser graph generally improves recall but reduces query throughput, while a sparser graph has the opposite effect. Finding the right balance requires testing multiple configurations, yet there is limited guidance on how to do this efficiently.  
+
+### Key HNSW hyperparameters  
+
+The three most important hyperparameters in HNSW are:  
+
+- **`M`** – The maximum number of graph edges per vector. Higher values increase memory usage but may improve search quality.  
+- **`efSearch`** – The size of the candidate queue during search. Larger values may improve search quality but increase search latency.  
+- **`efConstruction`** – Similar to `efSearch`, but used during index construction. Higher values improve search quality but increase index build time.  
+
+### Finding effective configurations  
+
+One approach to tuning these hyperparameters is **hyperparameter optimization (HPO)**, an automated technique that searches for the optimal configuration of a black-box function [5, 6]. However, HPO can be computationally expensive while providing limited benefits [3], especially in cases where the underlying algorithm is well understood.  
+
+An alternative is **transfer learning**, where knowledge from optimizing one dataset is applied to another. This approach helps identify configurations that approximate optimal results while maintaining efficiency [3, 4].  
+
+### Recommended HNSW configurations  
+
+In the next section, we introduce a method for selecting HNSW configurations. Based on this approach, we provide **five precomputed configurations** that progressively increase graph density. These configurations cover a range of trade-offs between search quality and speed across different datasets.  
+
+To optimize your search performance, you can **evaluate these five configurations sequentially**, stopping once recall meets your requirements. Because the configurations are ordered by increasing search quality, testing them in this order is likely to yield better search quality with each step:  
+
+```python
+{'M': 16,  'efConstruction': 128, 'efSearch': 32}
+{'M': 32,  'efConstruction': 128, 'efSearch': 32}
+{'M': 16,  'efConstruction': 128, 'efSearch': 128}
+{'M': 64,  'efConstruction': 128, 'efSearch': 128}
+{'M': 128, 'efConstruction': 256, 'efSearch': 256}
+```
+
+## Portfolio learning for HNSW  
+
+Portfolio learning [2, 3, 4] selects a set of complementary configurations so that at least one performs well on average when evaluated across different scenarios. Applying this approach to HNSW, we aim to identify a set of configurations that balance recall and query throughput.  
```

Comment:
```suggestion
Portfolio learning [2, 3, 4] selects a set of complementary configurations so that at least one performs well on average when evaluated across different scenarios. Applying this approach to HNSW, we aimed to identify a set of configurations that balance recall and query throughput.  
```